src/v/archival/types.cc

```diff
@@ -143,7 +143,9 @@ bool adjacent_segment_run::maybe_add_segment(
                 ntp, s));
         }
     } else {
-        if (meta.size_bytes + s.size_bytes <= max_size) {
+        if (
+          meta.size_bytes + s.size_bytes <= max_size
+          && s.is_compacted == meta.is_compacted) {
```

Comment:
Can you remind me if it's possible for local retention to GC non-compacted segments? If so, maybe if it's still worthwhile to merge as long as the offset range is below the `last_uploaded_compacted_offset`. Is there a downside to merging mismatched kinds of segments?

Reply:
We want to prevent it from racing with the compacted reuploader.

Reply:
Wouldn't a race not be possible as long as we merge only below `last_uploaded_compacted_offset`? If so, we could still e.g. merge segments that weren't reuploaded because of local GC